api/application/services/dataset_service.py

```diff
@@ -60,47 +65,65 @@ def _extract_sensitivities_and_domains(
                 sensitivities.update(sensitivities_dict.get(permission))
         return {"protected_domains": protected_domains, "sensitivities": sensitivities}
 
-    def _fetch_datasets(self, sensitivities_and_domains: Dict[str, Set[str]]):
-        authorised_datasets = set()
+    def _fetch_datasets(
+        self, sensitivities_and_domains: Dict[str, Set[str]], tag_filters
+    ):
+        authorised_datasets = list()
+
         if len(sensitivities_and_domains.get("sensitivities")) > 0:
             self._extract_datasets_from_sensitivities(
-                authorised_datasets, sensitivities_and_domains
+                authorised_datasets, sensitivities_and_domains, tag_filters
             )
         if len(sensitivities_and_domains.get("protected_domains")) > 0:
             self._extract_datasets_from_protected_domains(
-                authorised_datasets, sensitivities_and_domains
+                authorised_datasets, sensitivities_and_domains, tag_filters
             )
 
-        return sorted(authorised_datasets)
+        # Now filter the list to only get unique values
+        # return the values of a new dictionary that use the unique upload_path as a key
+        return sorted(
+            list(
+                {
+                    dataset.get_ui_upload_path(): dataset
+                    for dataset in authorised_datasets
+                }.values()
+            ),
+            key=lambda d: d.domain,
+        )
```

Comment:
This looks unnecessarily complex.

What was wrong with the code before that would get the unique values by converting the datasets to dataset.get_ui_upload_path() and then putting them in a set?